# **Determination of second-degree surface parameters (quadrics) by minimizing algebraic distances**
Author: Hubert Dębowski

In [ ]:
import numpy as np
import cmath
import math
from scipy.linalg import sqrtm
from scipy.optimize import minimize_scalar
from math import sqrt

Data:

In [ ]:
xyz = np.array([[39.600,42.911,0.001],
[39.606,42.918,33.542],
[30.064,58.441,16.995],
[39.402,76.961,0.006],
[39.403,76.969,50.002],
[60.266,77.161,16.996],
[60.269,77.169,50.004],
[69.724,63.322,0.007],
[69.727,63.320,33.542],
[39.606,42.911,16.994],
[39.600,42.918,50.003],
[30.065,58.441,0.002],
[30.062,58.448,33.542],
[30.060,58.442,50.004],
[39.403,76.967,16.994],
[39.409,76.961,33.542],
[60.261,77.169,0.003],
[60.267,77.163,33.541],
[69.720,63.323,16.995],
[69.724,63.328,50.002]])

print(xyz)

[[3.9600e+01 4.2911e+01 1.0000e-03]
 [3.9606e+01 4.2918e+01 3.3542e+01]
 [3.0064e+01 5.8441e+01 1.6995e+01]
 [3.9402e+01 7.6961e+01 6.0000e-03]
 [3.9403e+01 7.6969e+01 5.0002e+01]
 [6.0266e+01 7.7161e+01 1.6996e+01]
 [6.0269e+01 7.7169e+01 5.0004e+01]
 [6.9724e+01 6.3322e+01 7.0000e-03]
 [6.9727e+01 6.3320e+01 3.3542e+01]
 [3.9606e+01 4.2911e+01 1.6994e+01]
 [3.9600e+01 4.2918e+01 5.0003e+01]
 [3.0065e+01 5.8441e+01 2.0000e-03]
 [3.0062e+01 5.8448e+01 3.3542e+01]
 [3.0060e+01 5.8442e+01 5.0004e+01]
 [3.9403e+01 7.6967e+01 1.6994e+01]
 [3.9409e+01 7.6961e+01 3.3542e+01]
 [6.0261e+01 7.7169e+01 3.0000e-03]
 [6.0267e+01 7.7163e+01 3.3541e+01]
 [6.9720e+01 6.3323e+01 1.6995e+01]
 [6.9724e+01 6.3328e+01 5.0002e+01]]


In [ ]:
# extract the x, y, and z coordinates
x = xyz[:, 0]
y = xyz[:, 1]
z = xyz[:, 2]

# compute the parameters
x2 = x ** 2
y2 = y ** 2
z2 = z ** 2
xy = 2 * x * y
xz = 2 * x * z
yz = 2 * y * z
x2_ = 2 * x
y2_ = 2 * y
z2_ = 2 * z

# create the matrix with the parameters
A = np.vstack([x2, y2, z2, xy, xz, yz, x2_, y2_, z2_]).T

In [ ]:
print(A)

[[1.56816000e+03 1.84135392e+03 1.00000000e-06 3.39855120e+03
  7.92000000e-02 8.58220000e-02 7.92000000e+01 8.58220000e+01
  2.00000000e-03]
 [1.56863524e+03 1.84195472e+03 1.12506576e+03 3.39962062e+03
  2.65692890e+03 2.87911111e+03 7.92120000e+01 8.58360000e+01
  6.70840000e+01]
 [9.03844096e+02 3.41535048e+03 2.88830025e+02 3.51394045e+03
  1.02187536e+03 1.98640959e+03 6.01280000e+01 1.16882000e+02
  3.39900000e+01]
 [1.55251760e+03 5.92299552e+03 3.60000000e-05 6.06483464e+03
  4.72824000e-01 9.23532000e-01 7.88040000e+01 1.53922000e+02
  1.20000000e-02]
 [1.55259641e+03 5.92422696e+03 2.50020000e+03 6.06561901e+03
  3.94045761e+03 7.69720788e+03 7.88060000e+01 1.53938000e+02
  1.00004000e+02]
 [3.63199076e+03 5.95381992e+03 2.88864016e+02 9.30036965e+03
  2.04856187e+03 2.62285671e+03 1.20532000e+02 1.54322000e+02
  3.39920000e+01]
 [3.63235236e+03 5.95505456e+03 2.50040002e+03 9.30179692e+03
  6.02738215e+03 7.71751735e+03 1.20538000e+02 1.54338000e+02
  1.00008000e+02]
 [4.86

In [ ]:
A.shape


(20, 9)

In [ ]:
# create a 9-row matrix with -1 values
B = np.ones((20, 1)) * -1

# print the matrix
print(B)

[[-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]
 [-1.]]


In [ ]:
X = (np.matrix(A.T.dot(A)).I).dot(A.T).dot(B)
print(X)

[[ 1.75444448e-04]
 [ 1.75400627e-04]
 [-3.24200403e-08]
 [ 2.83658478e-08]
 [ 1.68191925e-09]
 [-1.97440177e-08]
 [-8.77434264e-03]
 [-1.05248017e-02]
 [ 1.87726499e-06]]


1.75444 * 10^-4
last 1.87726 * 10^-6

## 2. The coordinates of the center of symmetry

In [ ]:
MW = np.array([[X[0], X[3], X[4]],
              [X[3], X[1], X[5]],
              [X[4], X[5], X[2]]])
print(MW)

[[[[ 1.75444448e-04]]

  [[ 2.83658478e-08]]

  [[ 1.68191925e-09]]]


 [[[ 2.83658478e-08]]

  [[ 1.75400627e-04]]

  [[-1.97440177e-08]]]


 [[[ 1.68191925e-09]]

  [[-1.97440177e-08]]

  [[-3.24200403e-08]]]]


In [ ]:
MW = MW.reshape(3,3) #different reshape
MW = np.matrix(MW)
print(MW)

[[ 1.75444448e-04  2.83658478e-08  1.68191925e-09]
 [ 2.83658478e-08  1.75400627e-04 -1.97440177e-08]
 [ 1.68191925e-09 -1.97440177e-08 -3.24200403e-08]]


In [ ]:
W = X[6:9, 0]*-1 

In [ ]:
print(W)

[[ 8.77434264e-03]
 [ 1.05248017e-02]
 [-1.87726499e-06]]


In [ ]:
MW_1 = MW.I
print(MW_1)
MW_1.shape

[[ 5.69980720e+03 -8.88428232e-01  2.96241352e+02]
 [-8.88428232e-01  5.70084323e+03 -3.47189710e+03]
 [ 2.96241352e+02 -3.47189710e+03 -3.08429892e+07]]


(3, 3)

In [ ]:
S = MW_1.dot(W)
print(S)

[[50.00215468]
 [59.99896684]
 [23.95875848]]


## 3)	The dimensions of the object, i.e. the lengths of the semi-major axes of the quadric

In [ ]:
I1 = X[0]+X[1]+X[2]

I2 = np.linalg.det(np.array([[X[0], X[3]], [X[3], X[1]]]).reshape(2,2))+np.linalg.det((np.array([[X[1], X[5]],[X[5], X[2]]])).reshape(2,2))+np.linalg.det((np.array([[X[0], X[4]],[X[4], X[2]]])).reshape(2,2))

I3 = np.linalg.det(np.array([[X[0], X[3], X[4]],
               [X[3], X[1], X[5]],
               [X[4], X[5], X[2]]]).reshape(3,3))

In [ ]:
print(f'I1\n{I1}')
print(f'I2\n{I2}')
print(f'I3\n{I3}')

I1
[[0.00035081]]
I2
3.076169055643298e-08
I3
-9.977329106438352e-16


In [ ]:
p = np.subtract(I2, pow(I1,2)/3)
print(p)

[[-1.02614824e-08]]


In [ ]:
q = ((I1 * I2)/3) - ((2/27)*I1**3) - I3
print(q)

[[4.00094908e-13]]


In [ ]:
s = math.sqrt(-(p/3))
print(s)

5.8484990609966626e-05


In [ ]:
cosB = -(q/(2*s**3))
print(cosB)

[[-0.99999944]]


In [ ]:
beta = math.acos(cosB)
#beta = beta * 180/math.pi
print(beta)

3.1405311881711415


In [ ]:
lambda1 = (2*s*np.cos(beta/3))+I1/3
print(lambda1)

[[0.00017546]]


In [ ]:
lambda2 = ((2*s*np.cos((beta/3 + math.radians(120)))))+I1/3
print(lambda2)

[[-3.24222786e-08]]


In [ ]:
lambda3 = (2*s*np.cos(beta/3 + math.radians(240)))+I1/3
print(lambda3)

[[0.00017539]]


In [ ]:
d_arr = np.array([[X[0], X[3], X[4], X[6]],
                   [X[3], X[1], X[5], X[7]],
                   [X[4], X[5], X[2], X[8]],
                   [X[6], X[7], X[8], X[0]]]).reshape(4,4)
print(d_arr)

[[ 1.75444448e-04  2.83658478e-08  1.68191925e-09 -8.77434264e-03]
 [ 2.83658478e-08  1.75400627e-04 -1.97440177e-08 -1.05248017e-02]
 [ 1.68191925e-09 -1.97440177e-08 -3.24200403e-08  1.87726499e-06]
 [-8.77434264e-03 -1.05248017e-02  1.87726499e-06  1.75444448e-04]]


In [ ]:
BB= np.ones((16, 1)).reshape(4,4)
BB[3,3] = 0

print(BB)

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 0.]]


In [ ]:
z_arr = np.zeros((16, 1)).reshape(4,4)
z_arr[3,3] = 1
print(z_arr)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 1.]]


In [ ]:
p_arr = (d_arr*BB).reshape(4,4)
print(p_arr)

[[ 1.75444448e-04  2.83658478e-08  1.68191925e-09 -8.77434264e-03]
 [ 2.83658478e-08  1.75400627e-04 -1.97440177e-08 -1.05248017e-02]
 [ 1.68191925e-09 -1.97440177e-08 -3.24200403e-08  1.87726499e-06]
 [-8.77434264e-03 -1.05248017e-02  1.87726499e-06  0.00000000e+00]]


In [ ]:
I4 = np.linalg.det((z_arr + p_arr).reshape(4,4))
print(I4)

7.000921224078394e-17


In [ ]:
a2 = -I4/(lambda1*I3)
print(a2)

[[399.91415561]]


In [ ]:
b2 = -I4/(lambda2*I3)
print(b2)

[[-2164199.83712541]]


In [ ]:
c2 = -I4/(lambda3*I3)
print(c2)

[[400.07760743]]


In [ ]:
a=np.sqrt(a2)
print(a)

[[19.99785378]]


In [ ]:
b=cmath.sqrt(b2)
print(b)

1471.1219654146323j


In [ ]:
c=np.sqrt(c2)
print(c)

[[20.00194009]]


4) Inclination of the semi-axes a, b, c relative to the axes of the coordinate system in order x, y, z [degrees]:

axis A

In [ ]:
kA = np.array([[MW[0,1], MW[0,2]],
              [MW[1,1], MW[1,2]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[1,0] = lambda1
kA = np.linalg.det(kA - lambda_temp)
print(kA)

-4.629191866863179e-16


In [ ]:
lA = np.array([[MW[0,2], MW[0,0]],
              [MW[1,2], MW[1,0]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[0,1] = lambda1
lA = np.linalg.det(lA - lambda_temp)

print(lA)

-2.2737197817220507e-16


In [ ]:
mA = np.array([[MW[0,0], MW[0,1]],
              [MW[1,0], MW[1,1]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[1,1] = lambda1
lambda_temp[0,0] = lambda1
mA = np.linalg.det(mA - lambda_temp)
print(mA)

2.1152876262628332e-20


In [ ]:
cosAA = kA/(np.sqrt(kA**2+lA**2+mA**2))
angAA = math.acos(cosAA)
angAA = angAA * 180/math.pi
print(angAA)

153.8411177711553


In [ ]:
cosBA = lA/(np.sqrt(kA**2+lA**2+mA**2))
angBA = math.acos(cosBA)
angBA = angBA * 180/math.pi
print(angBA)

116.15888210706096


In [ ]:
cosGA = mA/(np.sqrt(kA**2+lA**2+mA**2))
angGA = math.acos(cosGA)
angGA = angGA * 180/math.pi
print(angGA)

89.99765005542852


axis B

In [ ]:
kA = np.array([[MW[0,1], MW[0,2]],
              [MW[1,1], MW[1,2]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[1,0] = lambda2
kA = np.linalg.det(kA - lambda_temp)

lA = np.array([[MW[0,2], MW[0,0]],
              [MW[1,2], MW[1,0]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[0,1] = lambda2
lA = np.linalg.det(lA - lambda_temp)

mA = np.array([[MW[0,0], MW[0,1]],
              [MW[1,0], MW[1,1]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[1,1] = lambda2
lambda_temp[0,0] = lambda2
mA = np.linalg.det(mA - lambda_temp)

cosAA = kA/(np.sqrt(kA**2+lA**2+mA**2))
angAA = math.acos(cosAA)
angAA = angAA * 180/math.pi
print(angAA)

cosBA = lA/(np.sqrt(kA**2+lA**2+mA**2))
angBA = math.acos(cosBA)
angBA = angBA * 180/math.pi
print(angBA)

cosGA = mA/(np.sqrt(kA**2+lA**2+mA**2))
angGA = math.acos(cosGA)
angGA = angGA * 180/math.pi
print(angGA)

90.00055021375881
89.99355158852275
0.006471842573395741


axis C

In [ ]:
kA = np.array([[MW[0,1], MW[0,2]],
              [MW[1,1], MW[1,2]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[1,0] = lambda3
kA = np.linalg.det(kA - lambda_temp)

lA = np.array([[MW[0,2], MW[0,0]],
              [MW[1,2], MW[1,0]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[0,1] = lambda3
lA = np.linalg.det(lA - lambda_temp)

mA = np.array([[MW[0,0], MW[0,1]],
              [MW[1,0], MW[1,1]]])
lambda_temp = np.zeros((4, 1)).reshape(2,2)
lambda_temp[1,1] = lambda3
lambda_temp[0,0] = lambda3
mA = np.linalg.det(mA - lambda_temp)

cosAA = kA/(np.sqrt(kA**2+lA**2+mA**2))
angAA = math.acos(cosAA)
angAA = angAA * 180/math.pi
print(angAA)

cosBA = lA/(np.sqrt(kA**2+lA**2+mA**2))
angBA = math.acos(cosBA)
angBA = angBA * 180/math.pi
print(angBA)

cosGA = mA/(np.sqrt(kA**2+lA**2+mA**2))
angGA = math.acos(cosGA)
angGA = angGA * 180/math.pi
print(angGA)

116.15888194199175
26.158882743905988
90.00603013316633


**Deviations from the coating (geometric distances)**

In [ ]:
A11 = p_arr[0][0]
A12 = p_arr[0][1]
A13 = p_arr[0][2]
A14 = p_arr[0][3]
A21 = p_arr[1][0]
A22 = p_arr[1][1]
A23 = p_arr[1][2]
A24 = p_arr[1][3]
A31 = p_arr[2][0]
A32 = p_arr[2][1]
A33 = p_arr[2][2]
A34 = p_arr[2][3]
A41 = p_arr[3][0]
A42 = p_arr[3][1]
A43 = p_arr[3][2]
A44 = p_arr[3][3]

In [ ]:
G = A11*x**2 + A22*y**2 + A33*z**2 + 2*A12*x*y + 2*A13*x*z + 2*A23*y*z + 2*A41*x + 2*A42*y + 2*A43*z + 1

In [ ]:
Nx = 2 * (A11*x + A12*y + A13*z + A14)
Nx = np.array(Nx).reshape(20,1)
print(Nx)

[[-0.00365105]
 [-0.00364883]
 [-0.00699619]
 [-0.00371859]
 [-0.00371808]
 [ 0.00360242]
 [ 0.00360358]
 [ 0.00692028]
 [ 0.00692145]
 [-0.00364889]
 [-0.00365088]
 [-0.0069959 ]
 [-0.00699683]
 [-0.00699748]
 [-0.00371819]
 [-0.00371603]
 [ 0.00360061]
 [ 0.00360283]
 [ 0.00691894]
 [ 0.00692045]]


In [ ]:
Ny = 2 * (A21*x + A22*y + A23*z + A24)
Ny = np.array(Ny).reshape(20,1)
print(Ny)

[[-0.00599412]
 [-0.00599299]
 [-0.00054739]
 [ 0.00595065]
 [ 0.00595148]
 [ 0.00602132]
 [ 0.00602282]
 [ 0.00116779]
 [ 0.00116576]
 [-0.00599479]
 [-0.00599364]
 [-0.00054672]
 [-0.00054559]
 [-0.00054835]
 [ 0.00595208]
 [ 0.00594932]
 [ 0.0060248 ]
 [ 0.00602137]
 [ 0.00116747]
 [ 0.00116792]]


In [ ]:
Nz = 2 * (A31*x + A32*y + A33*z + A34)
Nz = np.array(Nz).reshape(20,1)
print(Nz)

[[ 2.19320206e-06]
 [ 1.81446856e-08]
 [ 4.45982979e-07]
 [ 8.47644217e-07]
 [-2.39441300e-06]
 [-1.91703234e-07]
 [-2.33225043e-06]
 [ 1.48815500e-06]
 [-6.86168034e-07]
 [ 1.09139476e-06]
 [-1.04920806e-06]
 [ 1.54781383e-06]
 [-6.27208979e-07]
 [-1.69437619e-06]
 [-2.54092637e-07]
 [-1.32680918e-06]
 [ 9.09791533e-07]
 [-1.26455798e-06]
 [ 3.86598770e-07]
 [-1.75376176e-06]]


In [ ]:
a = A11*Nx**2 + A22*Ny**2 + A33*Nz**2+2*(A12*Nx*Ny + A13*Nx*Nz + A23*Ny*Nz)
print(a)

[[8.64200782e-09]
 [8.63678599e-09]
 [8.64019370e-09]
 [8.63575228e-09]
 [8.63681263e-09]
 [8.63742484e-09]
 [8.64207332e-09]
 [8.64175147e-09]
 [8.64375171e-09]
 [8.64064774e-09]
 [8.64077930e-09]
 [8.63934358e-09]
 [8.64143277e-09]
 [8.64355040e-09]
 [8.63821348e-09]
 [8.62963974e-09]
 [8.64248273e-09]
 [8.63804138e-09]
 [8.63835114e-09]
 [8.64221432e-09]]


In [ ]:
b = Nx**2+Ny**2+Nz**2
print(b)

[[4.92597011e-05]
 [4.92299379e-05]
 [4.92462982e-05]
 [4.92381480e-05]
 [4.92441947e-05]
 [4.92337210e-05]
 [4.92602196e-05]
 [4.92540703e-05]
 [4.92654722e-05]
 [4.92519522e-05]
 [4.92526987e-05]
 [4.92414563e-05]
 [4.92533643e-05]
 [4.92654313e-05]
 [4.92521793e-05]
 [4.92032952e-05]
 [4.92625622e-05]
 [4.92372347e-05]
 [4.92346879e-05]
 [4.92567086e-05]]


In [ ]:
c = A11*x**2 + A22*y**2 + A33*z**2 + 2*A12*x*y + 2*A13*x*z + 2*A23*y*z + 2*A14*x + 2*A24*y + 2*A34*z + 1
c = c.reshape(20,1)
print(c)

[[ 8.53286066e-06]
 [-1.82363063e-05]
 [ 2.11525492e-06]
 [-2.06567649e-06]
 [ 3.15836724e-06]
 [-1.12374787e-05]
 [ 6.09276747e-06]
 [-5.42137790e-06]
 [ 2.64549880e-05]
 [ 1.45406215e-05]
 [-4.82333140e-06]
 [-2.18210817e-05]
 [ 1.07834624e-05]
 [ 8.95062044e-06]
 [ 3.49657440e-05]
 [-3.61214870e-05]
 [ 1.28381826e-05]
 [-7.63908840e-06]
 [-1.60080358e-05]
 [-5.05375777e-06]]


In [ ]:
delta = b**2 - 4*a*c
print(delta)

[[2.42622319e-09]
 [2.42421680e-09]
 [2.42512478e-09]
 [2.42446657e-09]
 [2.42488160e-09]
 [2.42434754e-09]
 [2.42635862e-09]
 [2.42615085e-09]
 [2.42617207e-09]
 [2.42525223e-09]
 [2.42599504e-09]
 [2.42547510e-09]
 [2.42552116e-09]
 [2.42677326e-09]
 [2.42456900e-09]
 [2.42221112e-09]
 [2.42635622e-09]
 [2.42456923e-09]
 [2.42460763e-09]
 [2.42639805e-09]]


In [ ]:
t1 = ((-b**2) - np.sqrt(delta))/(2*a)
print(t1)

[[-2849.98200878]
 [-2850.52563743]
 [-2849.93501686]
 [-2851.0137478 ]
 [-2850.90773014]
 [-2850.39167811]
 [-2850.03994602]
 [-2850.02402664]
 [-2849.37703425]
 [-2849.86017161]
 [-2850.25315398]
 [-2850.42126755]
 [-2849.75926394]
 [-2849.79639346]
 [-2850.26172329]
 [-2851.70566124]
 [-2849.90353635]
 [-2850.31855551]
 [-2850.23890321]
 [-2850.01658145]]


In [ ]:
t2 = ((-b**2) + np.sqrt(delta))/(2*a)
print(t2)

[[2849.70122702]
 [2850.24502531]
 [2849.65432896]
 [2850.73300848]
 [2850.62695633]
 [2850.11104362]
 [2849.75916048]
 [2849.74330074]
 [2849.09624336]
 [2849.579434  ]
 [2849.97241213]
 [2850.14060722]
 [2849.47853574]
 [2849.5155965 ]
 [2849.98090398]
 [2851.42512064]
 [2849.6227374 ]
 [2850.037901  ]
 [2849.9582878 ]
 [2849.73584052]]


In [ ]:
sigma = np.sqrt((t2*Nx)**2 + (t2*Ny)**2 + (t2*Nz)**2)
print(sigma)

[[20.00070092]
 [19.9984732 ]
 [19.99765065]
 [20.00356487]
 [20.0040489 ]
 [19.99830148]
 [20.00121279]
 [19.99985304]
 [19.99762615]
 [19.99827298]
 [20.00118247]
 [20.00007987]
 [19.99785156]
 [20.00056125]
 [20.00113661]
 [20.00133877]
 [20.00073084]
 [19.99850184]
 [19.997426  ]
 [20.00033632]]



**Algebraic distances**

In [ ]:
def equation_loop( x_vals, y_vals, z_vals):
    result = []
    for i in range(len(x_vals)):
        x1, y1, z1 = x_vals[i], y_vals[i], z_vals[i]
        det = A11*A22*A33 + 2*A12*A13*A23 + A11*A23**2 + A22*A13**2 + A33*A12**2 - A11*A22*z1**2 - A22*A33*x1**2 - A33*A11*y1**2 - 2*A12*A23*x1*y1 - 2*A13*A23*y1*z1 - 2*A12*A13*x1*z1 + 2*A41*A23*z1 \
        + 2*A41*A13*y1 + 2*A42*A12*z1 + 2*A42*A23*x1 + 2*A43*A13*x1 + 2*A43*A12*y1
        result.append(det)
    return (result)

L=np.array(equation_loop(x,y,z)).reshape(20,1)
print(L)


[[ 3.45886121e-08]
 [-3.45955318e-05]
 [-8.85710043e-06]
 [ 5.66311935e-08]
 [-7.68947096e-05]
 [-8.81618956e-06]
 [-7.68801952e-05]
 [ 7.75686762e-08]
 [-3.45525550e-05]
 [-8.85680429e-06]
 [-7.69198367e-05]
 [ 3.53442436e-08]
 [-3.45947820e-05]
 [-7.69221648e-05]
 [-8.83475859e-06]
 [-3.45734892e-05]
 [ 7.73015183e-08]
 [-3.45507605e-05]
 [-8.81487680e-06]
 [-7.68737751e-05]]
